index.rst

```diff
@@ -216,6 +213,34 @@ Which of these is preferable probably depends on whether we want these operation
    TODO: think about how all of this affects disassembled composites.
    Or get rid of them!  I am pretty confident now that we'll never need them, and dropping that before we embark on other major changes to Datastore seems wise (sorry, Tim).
 
+.. _including-signed-urls-for-access-control:
+
+Including signed URLs for access control
+----------------------------------------
+
+Our access control model for the official Rubin Observatory data repositories (see :cite:`DMTN-169`) is based on information stored in the Registry - collection names, whenever possible, and new naming conventions or new columns in contexts (e.g. dimension records or dataset type names) that are not associated with collections.
+Access to the associated files managed by a Datastore is mediated by signed URLs; once the server side of the remote Butler has determined that an API call is permitted (based on that the Registry-side information), it can generate one or more signed URLs to pass to Datastore that provide direct access to the controlled files.
+A seemingly natural place to include URL-signing in this proposal is inside the Datastore implementation, since only the Datastore knows where any URLs might exist in the opaque-to-Registry records it uses, and only the Datastore ever uses and kind of URLs.
+This approach has two major drawbacks, however:
+
+- It requires the Datastore to have a server component; in every other respect we can abstract the differences between a SQL-backed full Butler and a client/server full Butler via a different Registry implementation (see :ref:`public-interface-changes`).
+
+- Because the information used to determine whether a URL *should* be signed lives in the Registry, a Datastore server cannot easily perform this job.
+  It cannot trust information provided to it via the `DatasetRef` and opaque table records passed to it from Butler in terms of access control, since those come from a public http interface; instead it'd have to use a Registry of its own to re-obtain Datastore records and collection information for each UUID it is passed in order to determine whether to sign URLs embedded in the records.
```

Comment:
Unless the Registry provides a signature over the access control information (at least) that can be passed through but not independently generated by the external client.

Reply:
I'm not sure I grokked "signature over the access control information" - you mean something cryptographically derived from the access control information and the UUID that the server could verify without access to the registry?

Reply:
> Now that we want registry server to report the records the registry server has to return records that include signed urls.

That's exactly what this most recent commit does.  Before this commit the datastore would have had to have that internal registry; I just didn't realize it.

Reply:
"signature over the access control information" means something like hashing the critical information along with a shared secret between the registry and the datastore servers.  That allows information to be passed via the client without the client being able to change it.

Reply:
I've [updated](4f4083aaa6cba0280a60972274d8b1f2631b4b05) the text to reflect this possibility.  I haven't changed my overall recommendation because the original proposal still wins on reducing the number of server round-trips and complexity (though my complexity metric gives a lot more weight to http client/server interfaces than regular Python interface, and that may be a lack-of-familiarity bias).